### ISSUE TITLE
[napari-worfklows] Relabeling

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-tasks-core#138]
Branching from https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/31#issuecomment-1280852826

**Comment by jluethi** [comment]
Very good point! My first thoughts on this: If our relabeling logic is 100% robust, we could apply the same logic to image and dataframe separately. But it's very important that they are relabeled the same way. So if there is any possible doubt, would be great to e.g. calculate a relabeling map on the image and apply that transformation to the table afterwards.

Also, I guess it will require having a good understanding in the table which label belongs to which ROI. Though given that we get them indiviudally per ROI, we could avoid concatenating it into a big table until the relabeling has taken place, thus simplifying this issue?

Regarding how this is processed: Do we do the relabeling once we have all the labels calculated or chunk-by-chunk? We could potentially save each chunk as it comes in, just relabeling it before saving and that should be stable, but if we need to do it only once everything is processed, that should also work

**Comment by tcompa** [comment]
Also for relabeling, like for other napari-workflows features, we probably need to restrict the scope of this task. See upcoming issue for broader discussion.

We 
1. The workflow only does labeling (goes from images to labels). In this case, we simply need a counter that keeps track of how many labels were found in the **previously-processed** ROIs, so that we can shift the labels of the **current** ROI by that amount.
2. The workflow only does measurement (from image/labels to dataframe). In this case, we do not need any labeling, because we assume that the input labels are already unique (across the different ROIs).
3. The workflow does both labeling and measurement on the just-created labels. In this case, we need to shift both the label array and the `label` column of the measurement dataframe.

Keeping track of the total number of labels and shifting things accordingly is simple, but we (me and @mfranzon) noticed that the choices to do within the task are becoming a bit complex. For instance: when there is a dataframe output, the decision on whether it needs to be relabeled or not depends on whether its input labels are coming from within the same workflow or from a pre-computed array of labels.

We don't have a quick fix for this, apart from asking for more info in the `output_spec` items. For instance each `label` output and each `dataframe` output may have a `relabel=True/False` item, and the setup would be up to the user.

Or perhaps we can add constraints on the supported workflows, and reach some confidence that we are covering all possible cases. We are not 100% happy with the current situation of the wrapper task, but let's continue in another issue.

**Comment by jluethi** [comment]
Very good discussion. Let's see if we can find good answers without exploding the scope.

> Keeping track of the total number of labels and shifting things accordingly is simple, but we (me and @mfranzon) noticed that the choices to do within the task are becoming a bit complex. For instance: when there is a dataframe output, the decision on whether it needs to be relabeled or not depends on whether its input labels are coming from within the same workflow or from a pre-computed array of labels.

Yes, good point. Relabeling could be an optional thing that raises a warning when there are non-unique labels in a dataframe, but it's harder to know a good default for label images based on their content.

> We don't have a quick fix for this, apart from asking for more info in the output_spec items. For instance each label output and each dataframe output may have a relabel=True/False item, and the setup would be up to the user.

Yes, seems like a reasonable solution to the problem to me. We're adding more and more to the output spec, which does make things a bit more complicated. But I do think the problem is inherently somewhat complicated, so I think that's fine.

------

Two more questions that came up for me when thinking about this:
On the fly relabeling vs. a consolidation function
Is relabeling running on-the-fly, i.e. whenever a ROI is processed, it's relabeled before it is saved? Or are all ROIs processed, labels are saved and then a "consolidation" step is run at the end of the workflow, which does the relabeling?

Relating a measurement to a label image?
I like the list of cases and agree that 1 & 2 are somewhat simple and important to have.
For case 3), when we need to relabel multiple things: Do we have a way to relate the relabeling we are doing? Because in theory, a user could label 2 things (i.e. organoids, maybe 20 in a well, and cells, maybe 50 per organoid) and make measurements for both. In that case, are we relabeling all things correctly?
Let's continue the discussion on this in https://github.com/fractal-analytics-platform/fractal-tasks-core/issues/174

**Comment by jluethi** [comment]
Actually, just thought about this a bit more, so to summarize: On the 3 scenarios you listed above, relabeling should be:
1) Relabel of the label image only
2) No relabeling
3) Relabeling of the label image and all dataframes with the same label mapping that was used for the label image relabeling

**Comment by tcompa** [comment]
> For case 3), when we need to relabel multiple things: Do we have a way to relate the relabeling we are doing? Because in theory, a user could label 2 things (i.e. organoids, maybe 20 in a well, and cells, maybe 50 per organoid) and make measurements for both. In that case, are we relabeling all things correctly?

Yes, the total counter of labels is specific for each (label or dataframe) output.

**Comment by tcompa** [comment]
As of our call with @jluethi a few minutes ago.

First, let's make `relabeling=True/False` a global task argument, not specific to any I/O item.

Second, let's handle the three cases above like this:
1. Act according to `relabeling` (defaulting to `True`).
2. Set `relabeling=False`, and issue a warning if it was `True`.
3. See point 1.

Third, let's move the `validate_labels_and_measurements` from the tests to the task, after some minor changes.

**Comment by tcompa** [comment]
One more question popped up:

We evoked the case where a certain measurement function does not output a row for **each** label, but skips some of them. For instance there labels `{1,2,..,20}`, and the measurement function outputs a dataframe including only rows for labels `{1,2,..,19}`.

I think this specific case escaped our reasoning.

Labels here are consecutive (from 1 to 19), and thus we cannot identify/handle anything weird during processing.

We can always validate the output at the end, but asserting that the same labels exist both in the dataframe and in the label array is not correct, because we already know that label 20 is missing from the dataframe (and rightly so, because the measurement function did not consider it to be part of the output - for whatever reason).

Still, we don't know how to relabel this dataframe in a self-consistent way (i.e. without resorting to the number of labels that were computed somewhere else).

The only way out (that I see) is that we do the following (for the three cases above):
1. No problem, act according to `relabeling` (defaulting to `True`).
2. No problem, set `relabeling=False` and proceed with whatever labels you received (issue a warning at the end, if the input labels were not unique).
3. Apart from acting accordingly to `relabeling`, we should handle differently the relabeling of label arrays (done as in point 1) and the relabeling of the measurement dataframes (which gets updated by the same number-of-current-labels from the label-array relabeling, rather than by its own number of labels).

**Comment by tcompa** [comment]
Some clarifications about previous statements

> > For case 3), when we need to relabel multiple things: Do we have a way to relate the relabeling we are doing? Because in theory, a user could label 2 things (i.e. organoids, maybe 20 in a well, and cells, maybe 50 per organoid) and make measurements for both. In that case, are we relabeling all things correctly?
> 
> Yes, the total counter of labels is specific for each (label or dataframe) output.

In view of the last comment: no, the cumulative label counters are not any more output-specific, because they are related to the only available label output (there has to be one and only one, if `relabeling=True`).

**Comment by tcompa** [comment]
I think the 3 usecases are now covered (as of b79a642f176c90a676a8318c9a600071302075c1). Missing points:
- [ ] Add a test for each use case.
- [ ] Move the output-validation function from tests to the task.

A last piece of the puzzle is missing (@jluethi, any opinion?): how can we validate the match of labels between a measurement table and a label array, if we accept that some measurement functions will not output rows for all of their input labels?
If this turns out to be impossible, let's keep the validation function in the test suite only (where we know that this edge case is not present).

**Comment by jluethi** [comment]
> In view of the last comment: no, the cumulative label counters are not any more output-specific, because they are related to the only available label output (there has to be one and only one, if relabeling=True)

Thus we do the same relabeling shifts of the labels for the label image and the dataframe labels, i.e. add the same offset to each for each ROI?

> how can we validate the match of labels between a measurement table and a label array, if we accept that some measurement functions will not output rows for all of their input labels?

This depends:
If the statement above is true, then missing label entries in a dataframe are fine, because the available measurements were shifted correctly and match the correct label in the label image.

If this match is not the case, then we need to raise an error here, because the output would potentially match the wrong things. In that case, I'd say we raise an error whenever labels in the label image & the dataframe don't 100% match.

**Comment by tcompa** [comment]
> > In view of the last comment: no, the cumulative label counters are not any more output-specific, because they are related to the only available label output (there has to be one and only one, if relabeling=True)
> 
> Thus we do the same relabeling shifts of the labels for the label image and the dataframe labels, i.e. add the same offset to each for each ROI?

Yes, that's my latest opinion. To be more explicit: There is a single cumulative number-of-labels counter (it is currently named `max_label_for_relabeling`), which is 0 at the beginning of the task and gets updated every time labeling is performed for a given ROI. When a measurement is performed, it uses `max_label_for_relabeling` for shifting labels but it does not update it.

> > how can we validate the match of labels between a measurement table and a label array, if we accept that some measurement functions will not output rows for all of their input labels?
> 
> This depends: If the statement above is true, then missing label entries in a dataframe are fine, because the available measurements were shifted correctly and match the correct label in the label image.

Indeed. In principle we are covered.
To be on the safe side we would need a test for this behavior. Is there any function (known to us) that applies a filter to labels before returning their measured properties? I shall open a new issue in a minute.

**Comment by tcompa** [comment]
I forgot to add that tests are now in-place for the three cases above, with the three (simplified) workflows `fractal-tasks-core/tests/data/napari_workflows/wf_relab_*.yaml`:
```YAML
# ==> wf_relab_1-labeling_only.yaml <==
!!python/object:napari_workflows._workflow.Workflow
_tasks:
  output_label: !!python/tuple
  - !!python/name:napari_segment_blobs_and_things_with_membranes.voronoi_otsu_labeling ''
  - input_image
  - 15.0
  - 2.0

# ==> wf_relab_2-measurement_only.yaml <==
  output_dataframe: !!python/tuple
  - !!python/name:napari_skimage_regionprops._regionprops.regionprops_table ''
  - input_image
  - input_label
  - true
  - true
  - false
  - false
  - false
  - false

# ==> wf_relab_3-labeling_and_measurement.yaml <==
  output_dataframe: !!python/tuple
  - !!python/name:napari_skimage_regionprops._regionprops.regionprops_table ''
  - input_image
  - output_label
  - true
  - true
  - false
  - false
  - false
  - false
```

**Comment by jluethi** [comment]
> Yes, that's my latest opinion. To be more explicit: There is a single cumulative number-of-labels counter (it is currently named max_label_for_relabeling), which is 0 at the beginning of the task and gets updated every time labeling is performed for a given ROI. When a measurement is performed, it uses max_label_for_relabeling for shifting labels but it does not update it.

Great, very nice that we got to such an implementation! Sounds good to me!
And very useful tests! I'll have to look around for a scenario 3 version where we only get a subset in the table, not aware of one just now (more likely to be an edge case when things go wrong or in custom scripts by users).

And great job finishing the relabeling that way! 👏🏻
